File: src/views/nodeCache/Trie.ts
// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT license.

import * as path from "path";
import { Uri } from "vscode";

export class Trie<T extends IUriData> {
    private _root: TrieNode<T>;

    constructor() {
        this._root = new TrieNode(null, null);
    }

    public insert(input: T): void {
        let currentNode: TrieNode<T> = this.root;
        const fsPath: string = Uri.parse(input.uri).fsPath;
        const segments: string[] = fsPath.split(path.sep);

        for (const segment of segments) {
            if (!segment) {
                continue;
            }
            if (!currentNode.children[segment]) {
                currentNode.children[segment] = new TrieNode(segment, null);
            }
            currentNode = currentNode.children[segment];
        }

        currentNode.value = input;
    }

    public find(fsPath: string): TrieNode<T> | undefined {
        let currentNode = this.root;
        const segments: string[] = fsPath.split(path.sep);

        for (const segment of segments) {
            if (!segment) {
                continue;
            }
            if (currentNode.children[segment]) {
                currentNode = currentNode.children[segment];
            } else {
                return undefined;
            }
        }

        return currentNode;
    }

    public findFirst(fsPath: string): TrieNode<T> | undefined {
        let currentNode = this.root;
        const segments: string[] = fsPath.split(path.sep);

        for (const segment of segments) {
            if (!segment) {
                continue;
            }
            if (currentNode.value) {
                return currentNode;
            }
            if (currentNode.children[segment]) {
                currentNode = currentNode.children[segment];
            } else {
                return undefined;
            }
        }

        return currentNode;
    }

    public findFirstAncestorNodeWithData(fsPath: string): TrieNode<T> | undefined {
        let currentNode: TrieNode<T> = this.root;
        let res: TrieNode<T> | undefined;
        const segments: string[] = fsPath.split(path.sep);

        for (const segment of segments) {
            if (!segment) {
                continue;
            }
            if (currentNode.children[segment]) {
                currentNode = currentNode.children[segment];
            } else {
                break;
            }

            if (currentNode.value) {
                res = currentNode;
            }
        }

        return res;
    }

    public get root(): TrieNode<T> {
        return this._root;
    }
}

export interface IUriData {
    uri: string;
}

export class TrieNode<T> {
    private _key: string;
    private _value: T;
    private _children: INodeChildren<T>;

    constructor(key: string, value: T) {
        this._key = key;
        this._value = value;
        this._children = {};
    }

    public get children(): INodeChildren<T> {
        return this._children;
    }

    public set children(children: INodeChildren<T>) {
        this._children = children;
    }

    public set value(value: T) {
        this._value = value;
    }

    public get value(): T | undefined {
        return this._value;
    }

}

interface INodeChildren<T> {
    [key: string]: TrieNode<T>;
}

File: jdtls.ext/com.microsoft.jdtls.ext.core/src/com/microsoft/jdtls/ext/core/model/Trie.java
/*******************************************************************************
 * Copyright (c) 2021 Microsoft Corporation and others.
 * All rights reserved. This program and the accompanying materials
 * are made available under the terms of the Eclipse Public License v1.0
 * which accompanies this distribution, and is available at
 * http://www.eclipse.org/legal/epl-v10.html
 *
 * Contributors:
 *     Microsoft Corporation - initial API and implementation
 *******************************************************************************/

package com.microsoft.jdtls.ext.core.model;

import java.util.HashSet;
import java.util.Map;
import java.util.Set;

import org.apache.commons.lang3.StringUtils;

public class Trie<T> {
    private TrieNode<T> root = new TrieNode<>();
    private Set<TrieNode<T>> allNodes = new HashSet<>();

    public Trie(Map<String, T> entries) {
        for (Map.Entry<String, T> entry : entries.entrySet()) {
            insert(entry.getKey(), entry.getValue());
        }
    }

    public Set<TrieNode<T>> getAllNodes() {
        return allNodes;
    }

    public void insert(String name, T value) {
        if (StringUtils.isBlank(name)) {
            // default package
            root.value = value;
            allNodes.add(root);
            return;
        }

        String[] names = name.split("\\.");
        TrieNode<T> currentNode = this.root;
        for (int i = 0; i < names.length; i++) {
            TrieNode<T> node;
            if (currentNode.children.containsKey(names[i])) {
                node = currentNode.children.get(names[i]);
            } else {
                node = new TrieNode<T>(names[i], null);
                currentNode.children.put(names[i], node);
                allNodes.add(node);
            }
            if (i == names.length - 1) {
                node.value = value;
            }

            currentNode = node;
        }
    }
}
